File: reference/k3x_ref/official_kda.py
# 공식 KDA의 BF16 경계와 V-first recurrent state를 검증하는 독립 scalar oracle입니다.
from __future__ import annotations

from dataclasses import dataclass

import torch
import torch.nn.functional as functional


@dataclass(frozen=True)
class OfficialKdaConfig:
    hidden_size: int
    heads: int
    head_dim: int
    conv_width: int
    rms_norm_epsilon: float
    gate_lower_bound: float

    @property
    def projection_size(self) -> int:
        return self.heads * self.head_dim


@dataclass(frozen=True)
class OfficialKdaWeights:
    q_proj: torch.Tensor
    k_proj: torch.Tensor
    v_proj: torch.Tensor
    q_conv: torch.Tensor
    k_conv: torch.Tensor
    v_conv: torch.Tensor
    f_a_proj: torch.Tensor
    f_b_proj: torch.Tensor
    a_log: torch.Tensor
    dt_bias: torch.Tensor
    b_proj: torch.Tensor
    g_proj: torch.Tensor
    o_norm: torch.Tensor
    o_proj: torch.Tensor


@dataclass(frozen=True)
class OfficialKdaState:
    conv_q: torch.Tensor
    conv_k: torch.Tensor
    conv_v: torch.Tensor
    recurrent_v_first: torch.Tensor


@dataclass(frozen=True)
class OfficialKdaBoundaries:
    projected_q: torch.Tensor
    projected_k: torch.Tensor
    projected_v: torch.Tensor
    convolved_q: torch.Tensor
    convolved_k: torch.Tensor
    convolved_v: torch.Tensor
    q: torch.Tensor
    k: torch.Tensor
    v: torch.Tensor
    log_decay: torch.Tensor
    beta: torch.Tensor
    recurrent_output: torch.Tensor
    gated: torch.Tensor


@dataclass(frozen=True)
class OfficialKdaResult:
    output: torch.Tensor
    state: OfficialKdaState
    boundaries: OfficialKdaBoundaries


def _invalid(message: str) -> ValueError:
    return ValueError(f"invalid official KDA: {message}")


def _require_tensor(
    name: str,
    tensor: torch.Tensor,
    shape: tuple[int, ...],
    dtype: torch.dtype,
    device: torch.device,
) -> None:
    if tensor.shape != shape:
        raise _invalid(f"{name} shape must be {shape}, got {tuple(tensor.shape)}")
    if tensor.dtype != dtype:
        raise _invalid(f"{name} dtype must be {dtype}, got {tensor.dtype}")
    if tensor.device != device:
        raise _invalid(f"{name} device must be {device}, got {tensor.device}")
    if not torch.isfinite(tensor).all():
        raise _invalid(f"{name} contains a non-finite value")


def _validate_config(config: OfficialKdaConfig) -> None:
    if min(config.hidden_size, config.heads, config.head_dim) <= 0:
        raise _invalid("hidden_size, heads, and head_dim must be positive")
    if config.conv_width < 2:
        raise _invalid("conv_width must be at least two")
    if config.rms_norm_epsilon <= 0:
        raise _invalid("rms_norm_epsilon must be positive")
    if config.gate_lower_bound >= 0:
        raise _invalid("gate_lower_bound must be negative")


def _validate_inputs(
    hidden: torch.Tensor,
    weights: OfficialKdaWeights,
    state: OfficialKdaState,
    config: OfficialKdaConfig,
) -> None:
    _validate_config(config)
    if hidden.ndim != 3:
        raise _invalid("hidden must have shape [batch, sequence, hidden_size]")
    batch, sequence, hidden_size = hidden.shape
    if batch <= 0 or sequence <= 0:
        raise _invalid("batch and sequence dimensions must be positive")
    if hidden_size != config.hidden_size:
        raise _invalid("hidden width does not match hidden_size")
    if hidden.dtype != torch.bfloat16:
        raise _invalid("hidden dtype must be torch.bfloat16")
    if not torch.isfinite(hidden).all():
        raise _invalid("hidden contains a non-finite value")

    device = hidden.device
    projection = config.projection_size
    bf16_shapes = {
        "q_proj": (projection, config.hidden_size),
        "k_proj": (projection, config.hidden_size),
        "v_proj": (projection, config.hidden_size),
        "f_a_proj": (config.head_dim, config.hidden_size),
        "f_b_proj": (projection, config.head_dim),
        "b_proj": (config.heads, config.hidden_size),
        "g_proj": (projection, config.hidden_size),
        "o_proj": (config.hidden_size, projection),
    }
    for name, shape in bf16_shapes.items():
        _require_tensor(name, getattr(weights, name), shape, torch.bfloat16, device)

    fp32_shapes = {
        "q_conv": (projection, config.conv_width),
        "k_conv": (projection, config.conv_width),
        "v_conv": (projection, config.conv_width),
        "a_log": (config.head_dim,),
        "dt_bias": (projection,),
        "o_norm": (config.head_dim,),
    }
    for name, shape in fp32_shapes.items():
        _require_tensor(name, getattr(weights, name), shape, torch.float32, device)

    history_shape = (batch, config.conv_width - 1, projection)
    _require_tensor("conv_q", state.conv_q, history_shape, torch.bfloat16, device)
    _require_tensor("conv_k", state.conv_k, history_shape, torch.bfloat16, device)
    _require_tensor("conv_v", state.conv_v, history_shape, torch.bfloat16, device)
    _require_tensor(
        "recurrent_v_first",
        state.recurrent_v_first,
        (batch, config.heads, config.head_dim, config.head_dim),
        torch.float32,
        device,
    )


def zero_official_kda_state(
    config: OfficialKdaConfig,
    batch_size: int,
    device: torch.device,
) -> OfficialKdaState:
    _validate_config(config)
    if batch_size <= 0:
        raise _invalid("batch_size must be positive")
    history_shape = (
        batch_size,
        config.conv_width - 1,
        config.projection_size,
    )
    recurrent_shape = (
        batch_size,
        config.heads,
        config.head_dim,
        config.head_dim,
    )
    return OfficialKdaState(
        conv_q=torch.zeros(history_shape, dtype=torch.bfloat16, device=device),
        conv_k=torch.zeros(history_shape, dtype=torch.bfloat16, device=device),
        conv_v=torch.zeros(history_shape, dtype=torch.bfloat16, device=device),
        recurrent_v_first=torch.zeros(
            recurrent_shape, dtype=torch.float32, device=device
        ),
    )


def _project(hidden: torch.Tensor, weight: torch.Tensor) -> torch.Tensor:
    return functional.linear(hidden, weight).to(torch.bfloat16)


def _short_conv(
    projected: torch.Tensor,
    history: torch.Tensor,
    weight: torch.Tensor,
) -> tuple[torch.Tensor, torch.Tensor]:
    outputs: list[torch.Tensor] = []
    current = history
    transposed_weight = weight.transpose(0, 1)
    for index in range(projected.shape[1]):
        window = torch.cat((current, projected[:, index : index + 1]), dim=1)
        convolved = (window.float() * transposed_weight.unsqueeze(0)).sum(dim=1)
        outputs.append(functional.silu(convolved).to(torch.bfloat16))
        current = window[:, 1:].to(torch.bfloat16)
    return torch.stack(outputs, dim=1), current


def _normalize_heads(value: torch.Tensor, scale: float = 1.0) -> torch.Tensor:
    normalized = functional.normalize(value.float(), p=2.0, dim=-1) * scale
    return normalized.to(torch.bfloat16)


def official_kda(
    hidden: torch.Tensor,
    weights: OfficialKdaWeights,
    state: OfficialKdaState,
    config: OfficialKdaConfig,
) -> OfficialKdaResult:
    _validate_inputs(hidden, weights, state, config)
    batch, sequence, _ = hidden.shape
    projection = config.projection_size
    head_shape = (batch, sequence, config.heads, config.head_dim)

    projected_q = _project(hidden, weights.q_proj)
    projected_k = _project(hidden, weights.k_proj)
    projected_v = _project(hidden, weights.v_proj)
    convolved_q, conv_q = _short_conv(projected_q, state.conv_q, weights.q_conv)
    convolved_k, conv_k = _short_conv(projected_k, state.conv_k, weights.k_conv)
    convolved_v, conv_v = _short_conv(projected_v, state.conv_v, weights.v_conv)

    q = _normalize_heads(
        convolved_q.reshape(head_shape), config.head_dim**-0.5
    )
    k = _normalize_heads(convolved_k.reshape(head_shape))
    v = convolved_v.reshape(head_shape)

    forget_low_rank = _project(hidden, weights.f_a_proj)
    forget = _project(forget_low_rank, weights.f_b_proj).reshape(head_shape)
    decay_argument = forget.float() + weights.dt_bias.reshape(
        1, 1, config.heads, config.head_dim
    )
    log_decay = config.gate_lower_bound * torch.sigmoid(
        weights.a_log.exp().reshape(1, 1, 1, config.head_dim) * decay_argument
    )
    beta = torch.sigmoid(_project(hidden, weights.b_proj).float())

    current_kv = state.recurrent_v_first.transpose(-1, -2).clone()
    recurrent_outputs: list[torch.Tensor] = []
    for index in range(sequence):
        alpha = log_decay[:, index].exp()
        decayed = alpha.unsqueeze(-1) * current_kv
        key = k[:, index].float()
        value = v[:, index].float()
        prediction = torch.einsum("bhk,bhkv->bhv", key, decayed)
        delta = (value - prediction) * beta[:, index].unsqueeze(-1)
        current_kv = decayed + key.unsqueeze(-1) * delta.unsqueeze(-2)
        recurrent_outputs.append(
            torch.einsum("bhk,bhkv->bhv", q[:, index].float(), current_kv)
        )
    recurrent_output = torch.stack(recurrent_outputs, dim=1)

    mean_square = recurrent_output.square().mean(dim=-1, keepdim=True)
    normalized = recurrent_output * torch.rsqrt(
        mean_square + config.rms_norm_epsilon
    )
    normalized = normalized * weights.o_norm.reshape(1, 1, 1, config.head_dim)
    output_gate = torch.sigmoid(
        _project(hidden, weights.g_proj).reshape(head_shape).float()
    )
    gated = (normalized * output_gate).to(torch.bfloat16)
    output = _project(gated.reshape(batch, sequence, projection), weights.o_proj)

    final_state = OfficialKdaState(
        conv_q=conv_q,
        conv_k=conv_k,
        conv_v=conv_v,
        recurrent_v_first=current_kv.transpose(-1, -2),
    )
    boundaries = OfficialKdaBoundaries(
        projected_q=projected_q,
        projected_k=projected_k,
        projected_v=projected_v,
        convolved_q=convolved_q,
        convolved_k=convolved_k,
        convolved_v=convolved_v,
        q=q,
        k=k,
        v=v,
        log_decay=log_decay,
        beta=beta,
        recurrent_output=recurrent_output,
        gated=gated,
    )
    return OfficialKdaResult(output=output, state=final_state, boundaries=boundaries)

File: runtime/src/official_kda.cpp
// 공식 KDA의 BF16 convolution과 FP32 V-first recurrence를 portable하게 실행합니다.
#include "k3x/official_kda.hpp"

#include "k3x/official_moe.hpp"

#include <algorithm>
#include <bit>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <utility>
#include <vector>

namespace k3x {
namespace {

bool product(std::size_t left, std::size_t right, std::size_t& output) {
    if (right && left > std::numeric_limits<std::size_t>::max() / right) {
        return false;
    }
    output = left * right;
    return true;
}

bool finite(std::span<const float> values) {
    return std::all_of(values.begin(), values.end(),
                       [](float value) { return std::isfinite(value); });
}

std::uint16_t encode_bf16(float value) noexcept {
    auto bits = std::bit_cast<std::uint32_t>(value);
    if ((bits & 0x7f800000U) == 0x7f800000U) {
        if (bits & 0x007fffffU) bits |= 0x00400000U;
        return static_cast<std::uint16_t>(bits >> 16U);
    }
    bits += 0x7fffU + ((bits >> 16U) & 1U);
    return static_cast<std::uint16_t>(bits >> 16U);
}

float rounded_bf16(float value) noexcept {
    return decode_bf16_word(encode_bf16(value));
}

bool valid_words(std::span<const std::uint16_t> values) {
    return std::all_of(values.begin(), values.end(), [](std::uint16_t value) {
        return std::isfinite(decode_bf16_word(value));
    });
}

bool valid_matrix(Bf16WeightView view, std::size_t rows, std::size_t cols) {
    std::size_t elements{};
    return view.rows == rows && view.cols == cols &&
           product(rows, cols, elements) && view.values.size() == elements &&
           valid_words(view.values);
}

Result<std::vector<float>> matmul_sequence(
    std::span<const float> input, std::size_t sequence,
    std::size_t input_width, Bf16WeightView weight) {
    std::size_t input_count{};
    if (!product(sequence, input_width, input_count) ||
        !valid_matrix(weight, weight.rows, input_width) ||
        input.size() != input_count || !finite(input)) {
        return Result<std::vector<float>>::failure(ErrorCode::invalid_extent);
    }
    std::size_t count{};
    if (!product(sequence, weight.rows, count)) {
        return Result<std::vector<float>>::failure(ErrorCode::invalid_extent);
    }
    std::vector<float> output(count);
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t row = 0; row < weight.rows; ++row) {
            double sum = 0.0;
            for (std::size_t column = 0; column < input_width; ++column) {
                sum += static_cast<double>(input[token * input_width + column]) *
                       decode_bf16_word(weight.values[row * input_width + column]);
            }
            const auto value = rounded_bf16(static_cast<float>(sum));
            if (!std::isfinite(value)) {
                return Result<std::vector<float>>::failure(
                    ErrorCode::invalid_extent);
            }
            output[token * weight.rows + row] = value;
        }
    }
    return Result<std::vector<float>>::success(std::move(output));
}

struct ConvResult {
    std::vector<float> output;
    std::vector<std::uint16_t> state;
};

Result<ConvResult> short_conv(
    std::span<const float> projected,
    std::span<const std::uint16_t> initial_state,
    std::span<const float> weight,
    std::size_t sequence,
    std::size_t projection,
    std::size_t width) {
    std::size_t history_count{};
    std::size_t weight_count{};
    std::size_t projected_count{};
    if (width < 2 || !product(width - 1, projection, history_count) ||
        !product(width, projection, weight_count) ||
        !product(sequence, projection, projected_count) ||
        initial_state.size() != history_count || weight.size() != weight_count ||
        projected.size() != projected_count || !valid_words(initial_state) ||
        !finite(weight)) {
        return Result<ConvResult>::failure(ErrorCode::invalid_extent);
    }
    std::vector<std::uint16_t> state(initial_state.begin(), initial_state.end());
    std::vector<float> output(sequence * projection);
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t channel = 0; channel < projection; ++channel) {
            float sum = 0.0F;
            for (std::size_t history = 0; history + 1 < width; ++history) {
                sum += decode_bf16_word(state[history * projection + channel]) *
                       weight[channel * width + history];
            }
            sum += projected[token * projection + channel] *
                   weight[channel * width + width - 1];
            const auto activated = sum / (1.0F + std::exp(-sum));
            const auto value = rounded_bf16(activated);
            if (!std::isfinite(value)) {
                return Result<ConvResult>::failure(ErrorCode::invalid_extent);
            }
            output[token * projection + channel] = value;
        }
        for (std::size_t history = 0; history + 2 < width; ++history) {
            for (std::size_t channel = 0; channel < projection; ++channel) {
                state[history * projection + channel] =
                    state[(history + 1) * projection + channel];
            }
        }
        for (std::size_t channel = 0; channel < projection; ++channel) {
            state[(width - 2) * projection + channel] =
                encode_bf16(projected[token * projection + channel]);
        }
    }
    return Result<ConvResult>::success({std::move(output), std::move(state)});
}

bool valid_config(const OfficialKdaConfig& config,
                  std::size_t& projection,
                  std::size_t& history,
                  std::size_t& recurrent) {
    std::size_t head_square{};
    return config.hidden_size && config.heads && config.head_dim &&
           config.conv_width >= 2 &&
           std::isfinite(config.rms_norm_epsilon) &&
           config.rms_norm_epsilon > 0.0F &&
           std::isfinite(config.gate_lower_bound) &&
           config.gate_lower_bound < 0.0F &&
           product(config.heads, config.head_dim, projection) &&
           product(config.conv_width - 1, projection, history) &&
           product(config.head_dim, config.head_dim, head_square) &&
           product(config.heads, head_square, recurrent);
}

bool valid_weights(const OfficialKdaWeightsView& weights,
                   const OfficialKdaConfig& config,
                   std::size_t projection) {
    std::size_t conv_count{};
    return product(projection, config.conv_width, conv_count) &&
           valid_matrix(weights.q_proj, projection, config.hidden_size) &&
           valid_matrix(weights.k_proj, projection, config.hidden_size) &&
           valid_matrix(weights.v_proj, projection, config.hidden_size) &&
           weights.q_conv.size() == conv_count && finite(weights.q_conv) &&
           weights.k_conv.size() == conv_count && finite(weights.k_conv) &&
           weights.v_conv.size() == conv_count && finite(weights.v_conv) &&
           valid_matrix(weights.f_a_proj, config.head_dim, config.hidden_size) &&
           valid_matrix(weights.f_b_proj, projection, config.head_dim) &&
           weights.a_log.size() == config.head_dim && finite(weights.a_log) &&
           weights.dt_bias.size() == projection && finite(weights.dt_bias) &&
           valid_matrix(weights.b_proj, config.heads, config.hidden_size) &&
           valid_matrix(weights.g_proj, projection, config.hidden_size) &&
           weights.o_norm.size() == config.head_dim && finite(weights.o_norm) &&
           valid_matrix(weights.o_proj, config.hidden_size, projection);
}

}  // namespace

OfficialKdaState zero_official_kda_state(const OfficialKdaConfig& config) {
    std::size_t projection{};
    std::size_t history{};
    std::size_t recurrent{};
    if (!valid_config(config, projection, history, recurrent)) return {};
    return {
        std::vector<std::uint16_t>(history),
        std::vector<std::uint16_t>(history),
        std::vector<std::uint16_t>(history),
        std::vector<float>(recurrent),
    };
}

Result<OfficialKdaResult> official_kda_cpu(
    std::span<const float> hidden,
    const OfficialKdaWeightsView& weights,
    const OfficialKdaState& state,
    const OfficialKdaConfig& config) {
    std::size_t projection{};
    std::size_t history_count{};
    std::size_t recurrent_count{};
    if (!valid_config(config, projection, history_count, recurrent_count) ||
        hidden.empty() || hidden.size() % config.hidden_size != 0 ||
        !finite(hidden) || !valid_weights(weights, config, projection) ||
        state.conv_q.size() != history_count ||
        state.conv_k.size() != history_count ||
        state.conv_v.size() != history_count ||
        state.recurrent_v_first.size() != recurrent_count ||
        !valid_words(state.conv_q) || !valid_words(state.conv_k) ||
        !valid_words(state.conv_v) || !finite(state.recurrent_v_first)) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }
    const auto sequence = hidden.size() / config.hidden_size;
    std::size_t sequence_projection{};
    if (!product(sequence, projection, sequence_projection)) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }
    std::vector<float> rounded_hidden(hidden.size());
    std::transform(hidden.begin(), hidden.end(), rounded_hidden.begin(),
                   [](float value) { return rounded_bf16(value); });
    if (!finite(rounded_hidden)) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }
    auto projected_q = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.q_proj);
    auto projected_k = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.k_proj);
    auto projected_v = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.v_proj);
    auto forget_low = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.f_a_proj);
    auto beta_projection = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.b_proj);
    auto output_gate = matmul_sequence(
        rounded_hidden, sequence, config.hidden_size, weights.g_proj);
    if (!projected_q || !projected_k || !projected_v || !forget_low ||
        !beta_projection || !output_gate) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }
    auto forget = matmul_sequence(
        forget_low.value(), sequence, config.head_dim, weights.f_b_proj);
    auto convolved_q = short_conv(
        projected_q.value(), state.conv_q, weights.q_conv,
        sequence, projection, config.conv_width);
    auto convolved_k = short_conv(
        projected_k.value(), state.conv_k, weights.k_conv,
        sequence, projection, config.conv_width);
    auto convolved_v = short_conv(
        projected_v.value(), state.conv_v, weights.v_conv,
        sequence, projection, config.conv_width);
    if (!forget || !convolved_q || !convolved_k || !convolved_v) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }

    std::vector<float> q(sequence_projection);
    std::vector<float> k(sequence_projection);
    const auto q_scale = 1.0F / std::sqrt(static_cast<float>(config.head_dim));
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t head = 0; head < config.heads; ++head) {
            const auto base = token * projection + head * config.head_dim;
            float q_squares = 0.0F;
            float k_squares = 0.0F;
            for (std::size_t channel = 0; channel < config.head_dim; ++channel) {
                const auto q_value = convolved_q.value().output[base + channel];
                const auto k_value = convolved_k.value().output[base + channel];
                q_squares += q_value * q_value;
                k_squares += k_value * k_value;
            }
            const auto q_inverse = 1.0F /
                std::max(std::sqrt(q_squares), 1.0e-12F);
            const auto k_inverse = 1.0F /
                std::max(std::sqrt(k_squares), 1.0e-12F);
            for (std::size_t channel = 0; channel < config.head_dim; ++channel) {
                q[base + channel] = rounded_bf16(
                    convolved_q.value().output[base + channel] *
                    q_inverse * q_scale);
                k[base + channel] = rounded_bf16(
                    convolved_k.value().output[base + channel] * k_inverse);
                if (!std::isfinite(q[base + channel]) ||
                    !std::isfinite(k[base + channel])) {
                    return Result<OfficialKdaResult>::failure(
                        ErrorCode::invalid_extent);
                }
            }
        }
    }

    std::vector<float> log_decay(sequence_projection);
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t head = 0; head < config.heads; ++head) {
            for (std::size_t channel = 0; channel < config.head_dim; ++channel) {
                const auto index = token * projection +
                                   head * config.head_dim + channel;
                const auto argument = std::exp(weights.a_log[channel]) *
                    (forget.value()[index] + weights.dt_bias[
                        head * config.head_dim + channel]);
                log_decay[index] = config.gate_lower_bound /
                    (1.0F + std::exp(-argument));
                if (!std::isfinite(argument) || !std::isfinite(log_decay[index])) {
                    return Result<OfficialKdaResult>::failure(
                        ErrorCode::invalid_extent);
                }
            }
        }
    }
    std::vector<float> beta(beta_projection.value().size());
    std::transform(beta_projection.value().begin(), beta_projection.value().end(),
                   beta.begin(), [](float value) {
                       return 1.0F / (1.0F + std::exp(-value));
                   });
    if (!finite(beta)) {
        return Result<OfficialKdaResult>::failure(ErrorCode::invalid_extent);
    }

    auto recurrent = state.recurrent_v_first;
    std::vector<float> recurrent_output(sequence_projection);
    std::vector<float> decayed(config.head_dim * config.head_dim);
    std::vector<float> prediction(config.head_dim);
    std::vector<float> delta(config.head_dim);
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t head = 0; head < config.heads; ++head) {
            const auto vector_base = token * projection + head * config.head_dim;
            const auto state_base = head * config.head_dim * config.head_dim;
            std::fill(prediction.begin(), prediction.end(), 0.0F);
            for (std::size_t key = 0; key < config.head_dim; ++key) {
                const auto alpha = std::exp(log_decay[vector_base + key]);
                for (std::size_t value = 0; value < config.head_dim; ++value) {
                    const auto state_index = state_base +
                        value * config.head_dim + key;
                    const auto kv_index = key * config.head_dim + value;
                    decayed[kv_index] = alpha * recurrent[state_index];
                    prediction[value] +=
                        k[vector_base + key] * decayed[kv_index];
                }
            }
            for (std::size_t value = 0; value < config.head_dim; ++value) {
                delta[value] =
                    (convolved_v.value().output[vector_base + value] -
                     prediction[value]) * beta[token * config.heads + head];
            }
            for (std::size_t key = 0; key < config.head_dim; ++key) {
                for (std::size_t value = 0; value < config.head_dim; ++value) {
                    const auto updated = decayed[key * config.head_dim + value] +
                        k[vector_base + key] * delta[value];
                    if (!std::isfinite(updated)) {
                        return Result<OfficialKdaResult>::failure(
                            ErrorCode::invalid_extent);
                    }
                    recurrent[state_base + value * config.head_dim + key] = updated;
                }
            }
            for (std::size_t value = 0; value < config.head_dim; ++value) {
                float output = 0.0F;
                for (std::size_t key = 0; key < config.head_dim; ++key) {
                    output += q[vector_base + key] *
                        recurrent[state_base + value * config.head_dim + key];
                }
                recurrent_output[vector_base + value] = output;
                if (!std::isfinite(output)) {
                    return Result<OfficialKdaResult>::failure(
                        ErrorCode::invalid_extent);
                }
            }
        }
    }

    std::vector<float> gated(sequence_projection);
    for (std::size_t token = 0; token < sequence; ++token) {
        for (std::size_t head = 0; head < config.heads; ++head) {
            const auto base = token * projection + head * config.head_dim;
            float squares = 0.0F;
            for (std::size_t value = 0; value < config.head_dim; ++value) {
                const auto item = recurrent_output[base + value];
                squares += item * item;
            }
            const auto inverse = 1.0F / std::sqrt(
                squares / static_cast<float>(config.head_dim) +
                config.rms_norm_epsilon);
            for (std::size_t value = 0; value < config.head_dim; ++value) {
                const auto gate = 1.0F /
                    (1.0F + std::exp(-output_gate.value()[base + value]));
                gated[base + value] = rounded_bf16(
                    recurrent_output[base + value] * inverse *
                    weights.o_norm[value] * gate);
                if (!std::isfinite(gated[base + value])) {
                    return Result<OfficialKdaResult>::failure(
                        ErrorCode::invalid_extent);
                }
            }
        }
    }
    auto output = matmul_sequence(
        gated, sequence, projection, weights.o_proj);
    if (!output) {
        return Result<OfficialKdaResult>::failure(
            output.error(), output.message());
    }
    OfficialKdaBoundaries boundaries{
        std::move(projected_q.value()),
        std::move(projected_k.value()),
        std::move(projected_v.value()),
        std::move(convolved_q.value().output),
        std::move(convolved_k.value().output),
        std::move(convolved_v.value().output),
        std::move(q),
        std::move(k),
        {},
        std::move(log_decay),
        std::move(beta),
        std::move(recurrent_output),
        std::move(gated),
    };
    boundaries.v = boundaries.convolved_v;
    OfficialKdaState final_state{
        std::move(convolved_q.value().state),
        std::move(convolved_k.value().state),
        std::move(convolved_v.value().state),
        std::move(recurrent),
    };
    return Result<OfficialKdaResult>::success({
        std::move(output.value()), std::move(final_state), std::move(boundaries)});
}

}  // namespace k3x
